### DISCUSSION TITLE
Program freezes if compiled in debug mode.

### DISCUSSION THREAD

**Discussion by crmoratelli** [Rahix/avr-hal#528] [Q&A] [answered]
Hi.

Recently I was working on this project [BMP180-ATmega2560-rust](https://github.com/crmoratelli/BMP180-ATmega2560-rust), where I implemented a driver to the BMP180 using this HAL. I was using _cargo run_ to upload code to an Arduino Mega 2560 board. 

Everything was working fine. However, as the code got bigger, the program started to freeze. I could still see initialization messages, but it stopped execution at some point. I did some tests, and I noticed that commenting out some code lines allows the execution to continue. However, the local variables had wrong values, which I checked with prints. 

Compiling with _cargo run --release_ makes it work without any problem. My experience in C programming with embedded makes me think that, for some reason, the stack could be corrupted. I would like to understand better if there is any restriction in building a debugging target (_cargo run_) or what else could be causing this behavior. 

The link in this message takes you to the repository. This error can be reproduced with _cargo run_ and the running version with _cargo run --release_, but you need a BMP180 to see it working.

**Comment by Rahix** [comment, marked as answer]
If, by _debug build_, you mean a build without optimizations, you are pretty much guaranteed to run into tons of issues.  Due to the extremely limited amount of SRAM on AVR microcontrollers, we must rely on a lot of compiler optimizations to get the code into a shape whose memory footprint is small enough.
Otherwise, the stack will quickly grow too large and either corrupt statics or simply lead to an overflow.

Because of this, we usually configure the _dev_ profile (that's `cargo run`) to also apply optimizations, similar to the _release_ profile (`cargo run --release`).  The relevant code is here:

https://github.com/Rahix/avr-hal/blob/2eb28fa3024ecbf9898c8e900e744d1201ceffca/Cargo.toml#L1-L11

Please make sure you are also using similar settings in your `Cargo.toml`.  If you are still encountering issues, some other difference between the two profiles must be at play.  I'd try adding one of the following settings to the `dev` profile to see whether one of them is to blame:

```toml
codegen-units = 1
incremental = false
debug-assertions = false
overflow-checks = false
```

**Reply by crmoratelli** [reply]
Thank you! Your response make things very clear. 

My original dev profile that came with _cargo generate_ was:

```
[profile.dev]
panic = "abort"
lto = true
opt-level = "s"
```

Adding _overflow-checks=false_ made the coding run. 
```
[profile.dev]
panic = "abort"
lto = true
opt-level = "s"
overflow-checks = false`
```
Now, I can run it with _cargo run_! However, I understood the need for optimization.

Thanks!

**Reply by Rahix** [reply]
Interesting to see `overflow-checks` being the cause here.  Do you do something particularly math heavy in your program?

Disabling `overflow-checks` isn't necessarily bad, just keep in mind the [implications](https://doc.rust-lang.org/reference/expressions/operator-expr.html#overflow).

**Reply by crmoratelli** [reply]
Yes! BMP180 is a I2C device capable to read pressure and temperature. Pressure calculation from raw data is math heavy. 

By _"operators will panic..."_ means a processor's exception, consequently making the program to stop? 

Maybe I had two problems: 1) corrupted statics because I could see wrong values in some variables, and; 2) overflows in math operations. 

As I am new to Rust, I am trying to understand the basics.

**Reply by Rahix** [reply]
A _panic_ is sort of a program-controlled abort.  So not a processor exception, but rather the program intentionally stopping because of an unrecoverable error.  Read more about this in the [Rust Book: Chapter 9.1](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html).

Now that you mention it, you could very well have seen panics due to overflow.  So be careful here...

In embedded rust, we have to explicitly define the panicking strategy.  The most simple way is using the [`panic-halt`](https://crates.io/crates/panic-halt) crate which simply halts the CPU on panic.  The line that pulls in `panic-halt` is this one:

https://github.com/Rahix/avr-hal/blob/2eb28fa3024ecbf9898c8e900e744d1201ceffca/examples/arduino-uno/src/bin/uno-blink.rs#L7

The disadvantage is of course that you don't get to see any indication of the panic or where it occurred.  So a more sophisticated alternative is to provide your own panic handler with custom logic.  For example, you could print the location to serial console and also blink an error LED:

https://github.com/Rahix/avr-hal/blob/2eb28fa3024ecbf9898c8e900e744d1201ceffca/examples/arduino-uno/src/bin/uno-panic.rs#L20-L53

**Reply by crmoratelli** [reply]
Nice! I understand the error handling much better now. This panic handler will save me much time in future developments. Thanks!
